### PR TITLE
Move merged PR job directories to `trash_bin_dir`

### PR DESCRIPTION
**Pull request by Neves-P**
This PR adds a handler function for merged PRs that then moves the job directories into a `trash_bin_dir`, the location of which can be set via the config.

At the moment there is no scripting or automation to delete the contents of `trash_bin_dir`, this functionality will follow later.

The next step is testing the current implementation on a bot instance not running in `software.eessi.io`:
- [x] Merge events on PRs are detected correctly
- [x] Directories are moved to the appropriate `trash_bin_root_dir` and have the correct subdirectory structure that makes `trash_bin_dir`.

Most likely the `config` files need to be adjusted and things are likely to fail as the new additions have not run within the context of a bot instance. I will do so on Hábrók in Groningen and will report back on progress here.

### CODE REVIEW

File: eessi_bot_event_handler.py [hunk 1]
```diff
@@ -599,6 +601,61 @@ def start(self, app, port=3000):
...
+        Returns:
+        github.IssueComment.IssueComment instance or None (note, github refers to
+        PyGithub, not the github from the internal connections module)
+        """
+
+        # Detect event and only act if PR is merged
+        request_body = event_info['raw_request_body']
+        action = request_body['action']
+
+        if action == 'merged':
```
Comment (trz42) [inline eessi_bot_event_handler.py:622]: Maybe we should verify if this would also apply to PRs that get closed?

File: eessi_bot_event_handler.py [hunk 2]
```diff
@@ -599,6 +601,61 @@ def start(self, app, port=3000):
...
+            self.log("PR merged: scanning directories used by PR")
+            self.log(f"merge '{action}' action is handled")
```
Comment (trz42) [inline eessi_bot_event_handler.py:624]:
```suggestion
            self.log(f"pull_request event with action '{action}' will be handled")
```

File: eessi_bot_event_handler.py [hunk 3]
```diff
@@ -599,6 +601,61 @@ def start(self, app, port=3000):
...
+        else:
+            self.log(f"merge action '{action}' not handled")
```
Comment (trz42) [inline eessi_bot_event_handler.py:626]:
```suggestion
            self.log(f"pull_request event action '{action}' is not handled")
```

File: eessi_bot_event_handler.py [hunk 4]
```diff
@@ -599,6 +601,61 @@ def start(self, app, port=3000):
...
+        clean_up_comments_cfg = self.cfg[config.SECTION_CLEAN_UP_COMMENTS]
+        # 1) determine the jobs that have been run for the PR
+        job_dirs = determine_job_dirs(pr.number)
+
+        # 2) Get trash_bin_dir from configs
+        trash_bin_root_dir = self.cfg[config.SECTION_MERGED_PR][config.MERGED_PR_SETTING_TRASH_BIN_ROOT_DIR]
+        repo_cfg = get_repo_cfg(self.cfg)
+        repo_name = repo_cfg[cvmfs_repository.REPOS_CFG_REPO_NAME]
+        dt = datetime.now(timezone.utc)
+        trash_bin_dir = "/".join([trash_bin_root_dir, repo_name, dt.strftime('%Y%m%d')])
```
Comment (trz42) [inline eessi_bot_event_handler.py:641]: I'm a bit unsure whether we want to use the CVMFS repository or rather the target repository to which the merged/closed PR was made (e.g., EESSI/software-layer). The target repository should be available from the `event_info` dictionary.

File: tasks/clean_up.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# EESSI software layer, see https://github.com/EESSI/software-layer
+#
+# author: Bob Droege (@bedroge)
+# author: Kenneth Hoste (@boegel)
+# author: Hafsa Naeem (@hafsa-naeem)
+# author: Jacob Ziemke (@jacobz137)
+# author: Jonas Qvigstad (@jonas-lq)
+# author: Lara Ramona Peeters (@laraPPr)
+# author: Pedro Santos Neves (@Neves-P)
+# author: Thomas Roeblitz (@trz42)
```
Comment (trz42) [inline tasks/clean_up.py:14]:
```suggestion
# author: Pedro Santos Neves (@Neves-P)
```

File: tasks/clean_up.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import sys
+
+# Third party imports (anything installed into the local Python environment)
+from pyghee.utils import log
+
+# Local application imports (anything from EESSI/eessi-bot-software-layer)
+from tools import run_cmd
+
+
+def move_to_trash_bin(trash_bin_dir, job_dirs):
```
Comment (trz42) [inline tasks/clean_up.py:29]: Could you add a short docstring?

File: tasks/clean_up.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    funcname = sys._getframe().f_code.co_name
+    log(f"{funcname}(): trash_bin_dir = {trash_bin_dir}")
+
+    move_cmd = ["mkdir -p trash_bin_dir && mv -t", trash_bin_dir]
+    for job_dir in job_dirs:
+        move_cmd.append(job_dir)
+        ' '.join(move_cmd)
+        out, err, ec = run_cmd(move_cmd, 'Move job directories to trash_bin', raise_on_error=False)
```
Comment (trz42) [inline tasks/clean_up.py:37]: It is not particularly wrong to use shell commands. However, there may be Python routines to create a directory and to move directories.

**Review by trz42 — CHANGES_REQUESTED** [review]
Had a look at the PR. Looks good overall. Some small suggestions here and there. I wonder if we would want to keep the `cleanup_pr.sh` script. It might allow us to clean up manually. We could remove it later if the whole auto-cleanup works smoothly.

File: eessi_bot_event_handler.py (hunk 3 above)
Comment (Neves-P) [reply eessi_bot_event_handler.py:626]: Resolved in d645202

File: tasks/clean_up.py (hunk 5 above)
Comment (Neves-P) [reply tasks/clean_up.py]: Resolved in d75bb11

File: eessi_bot_event_handler.py (hunk 1 above)
Comment (Neves-P) [reply eessi_bot_event_handler.py]: Handled on 435bcae, will test it now

File: tasks/clean_up.py (hunk 6 above)
Comment (Neves-P) [reply tasks/clean_up.py:23]: Resolved in 6b38f37

**Comment by Neves-P** [comment]
Made some progress, currently the code crashes with 

```
[20240611-T17:09:32] Event received (id: 9ac77d80-2804-11ef-91d2-e4f7a18cac40, type: pull_request, action: closed), event data logged at /home1/f115372/bot/eessi-bot-software-layer/events_log/pull_request/closed/2024-06-11/2024-06-11T15-09-32_9ac77d80-2804-11ef-91d2-e4f7a18cac40
[20240611-T17:09:32] Request verified: signature OK!
[20240611-T17:09:32] [event id 9ac77d80-2804-11ef-91d2-e4f7a18cac40] Handler found for event type 'pull_request' (action: closed)
[20240611-T17:09:33] get_build_env_cfg(): jobs_base_dir '/home1/f115372/jobs'
[20240611-T17:09:33] get_build_env_cfg(): local_tmp '/tmp/f115372/EESSI'
[20240611-T17:09:33] get_build_env_cfg(): build_job_script '/home1/f115372/bot/eessi-bot-software-layer/scripts/bot-build.slurm'
[20240611-T17:09:33] get_build_env_cfg(): submit_command '/usr/bin/sbatch'
[20240611-T17:09:33] get_build_env_cfg(): slurm_params '--hold --mem=8G --hold'
[20240611-T17:09:33] get_build_env_cfg(): shared_fs_path: '/scratch/f115372/bot_shared_directory'
[20240611-T17:09:33] get_build_env_cfg(): container_cachedir '/scratch/f115372/bot_container_cachedir'
[20240611-T17:09:33] get_build_env_cfg(): build_logs_dir '/scratch/f115372/bot_build_logs'
[20240611-T17:09:33] {fn}(): cvmfs_customizations '{cvmfs_customizations_str}'
[20240611-T17:09:33] get_build_env_cfg(): cvmfs_customizations '{}'
[20240611-T17:09:33] get_build_env_cfg(): http_proxy 'None'
[20240611-T17:09:33] get_build_env_cfg(): https_proxy 'None'
[20240611-T17:09:33] get_build_env_cfg(): load_modules 'EasyBuild/4.9.1'
[20240611-T17:09:33] determine_job_dirs(): jobs_base_dir = /home1/f115372/jobs
[20240611-T17:09:33] determine_job_dirs(): date_pr_job_pattern = [0-9][0-9][0-9][0-9].[0-9][0-9]/pr_30/[0-9]*
[20240611-T17:09:33] determine_job_dirs(): glob_str = /home1/f115372/jobs/[0-9][0-9][0-9][0-9].[0-9][0-9]/pr_30/[0-9]*
[20240611-T17:09:33] WARNING: A crash occurred!
Traceback (most recent call last):
  File "/home1/f115372/bot/venv_eessi_bot/lib/python3.11/site-packages/pyghee/lib.py", line 170, in process_event
    self.handle_event(event_info, log_file=log_file)
  File "/home1/f115372/bot/venv_eessi_bot/lib/python3.11/site-packages/pyghee/lib.py", line 102, in handle_event
    handler(event_info, log_file=log_file)
  File "/home1/f115372/bot/eessi-bot-software-layer/eessi_bot_event_handler.py", line 435, in handle_pull_request_event
    handler(event_info, pr)
  File "/home1/f115372/bot/eessi-bot-software-layer/eessi_bot_event_handler.py", line 641, in handle_pull_request_closed_event
    repo_name = event_info['repository']['full_name']
                ~~~~~~~~~~^^^^^^^^^^^^^^
KeyError: 'repository'
```

There is an error in getting the `repo_name` with `repo_name = event_info['repository']['full_name']`, but that should be a simple fix. I'll carry on tomorrow :)

File: tasks/clean_up.py (hunk 7 above)
Comment (Neves-P) [reply tasks/clean_up.py]: Resolved in 6b38f37

**Comment by Neves-P** [comment]
We are getting there, but there are still a few details to consider.
A first successful move of the job directory happens in this PR on the bot instance on Hábrók: https://github.com/Neves-Bot/software-layer/pull/31#issuecomment-2162445167

- The comment clearly needs to be fixed as it doesn't actually report anything useful
- The moved directory was: `/home1/f115372/jobs/2024.06/pr_31/9915462`. This is fine, but it means that `2024.06/pr_31/` is left behind, with `pr_31/` being an empty directory. I considered moving `/pr_31`, but that would be trickier when a PR is open during more than one month, since then we would have `XXXX.YY/pr_NN/jobID/` _and_ `XXXX.ZZ/pr_NN/jobID/` and would not be able to use `determine_job_dirs()` directly which is quite a neat solution. I can think of two options: 
  - We keep this approach and move the `jobID/` directories within a `month/` directory of a given PR without regard for keeping that directory structure and then delete the lingering empty directories as a last step
  - We traverse the `month/` directories for the right PR number and move them to a (most likely existing) `month/` directory in `trash_bin_dir`. This is doable, but might add too much complexity just for cleaning up files. With the previous approach we can always find the tarballs by looking up the jobID in the comments of the PRs in question.

File: eessi_bot_event_handler.py (hunk 4 above)
Comment (Neves-P) [reply eessi_bot_event_handler.py]: Handled in 769daec68918d09ee586dfac4795c0cc7719bced and 4160bd1

**Comment by Neves-P** [comment]
As it stands now the bot will:

- Move the contents of the `pr_XX` directory with `copytree(dirs_exist_ok = True)` to `trash_bin_dir`. This includes the `event_XXXX` directory(ies) that contain the cloned PRs of the bot where the `slurm.out` files and other metadata also get dumped
- Reports back which directories were moved in a comment containing the directory(ies) where the files were moved to (`trash_bin_dir/YYYY.MM.DD`)

I think this is ready for review. I have two questions to look into as part of the review: 

- Is there any case where `dirs_exist_ok = True` is problematic? I don't think so, as the PR numbers are unique and each PR will only be moved once, so setting this argument to `True` just makes it easier copying the entire directory tree without extra book-keeping, but maybe I'm wrong
- Will merging PRs that span multiple months still be copied correctly? I think so, but I'm going to test that by mimicking such a case and merging a PR to see what happens.

File: eessi_bot_event_handler.py (hunk 1 above)
Comment (Neves-P) [reply eessi_bot_event_handler.py]: > Handled on [435bcae](https://github.com/EESSI/eessi-bot-software-layer/commit/435bcae9e0f1027489d27470f0e4b6a01412055f), will test it now

Seems to work!

**Comment by Neves-P** [comment]
Confirmed moving directories happens even when the PR spans more than one month in https://github.com/Neves-Bot/software-layer/pull/47#issuecomment-2219781334

**Review by trz42 — APPROVED** [review]
Looks good to me. Thanks @Neves-P !

**Pull request merged**